File: python/saddle-points/saddle_points.py
def saddle_points(matrix):
    """
    Search for saddle points within the given matrix. Return them in a
    list where each item is a dictionary with row and column keys.
    Parameters:
    * matrix: A list composed of lists of numbers. Ex: [[1, 2], [3, 4]]
    Returns:
    * A list of dictionary objects with two items 'row', and 'columns' where
    the keys show the position of a saddle point. Ex: [{"row":3, "column": 5}]
    Note that upper left of the matrix is referenced as 1, 1, not 0, 0.
    """

    # Sanity Check
    if any(len(row) != len(matrix[0]) for row in matrix):
        raise ValueError('Irregular matrix detected.')
    
    results = []

    # Can't really iterate if we have an empty matrix. Fails a 
    # test case if we try.
    if len(matrix) > 0:
        rows = len(matrix)
        cols = len(matrix[0])        
        for row in range(rows):
            for col in range(cols):
                # Check if any item greater or equal to all in row.
                if any(matrix[row][i] > matrix[row][col] for i in range(cols) if i != col):
                    continue
                
                # Check if any item less or equal to all in column.
                if any(matrix[j][col] < matrix[row][col] for j in range(rows) if j != row):
                    continue
                results.append({"row": row + 1, "column": col + 1})
    
    # Add an empty dictionary for when we have no results, just to
    # appease the test cases.
    if len(results) == 0:
        results.append(dict())
    
    return results

File: php/saddle-points/SaddlePoints.php
<?php
// Saddle Point exercise.

declare(strict_types=1);

// Find the Saddle Points in a 2D matrix. A saddle point is one that is both smallest in its column and largest in its row.
// @param $matrix: The 2D matrix to search on.
// @returns: An array of points. Each point is an array with keys of "row", and "column"
function saddlePoints(array $matrix): array
{
    $results = array();
    $height = count($matrix);
    $width = 0;
    if ($height > 0) {
        $width = count($matrix[0]);
    }

    for($j = 0; $j < $height; $j++) {
        for($i = 0; $i < $width; $i++) {
            if (isSaddlePoint($matrix, $i, $j, $width, $height)) {
                $results[] = array(
                    "row" => $j + 1, 
                    "column" => $i + 1,
                );                
            }
        }
    }

    return $results;
}

// Check to see if a given point is a saddle point in the matrix
// @param $matrix: The matrix to lookup data from
// @param $x: x-coordinate of point to look at
// @param $y: y-coordinate of point to look at
// @param $width: How wide is the 2D matrix
// @param $height: How tall is the 2D matrix
// @returns: True if the given point on the matrix is a saddle point.
function isSaddlePoint(array $matrix, int $x, int $y, int $width, int $height) : bool {
    if ($x < 0 || $x >= $width || $y < 0 || $y >= $height) {
        return false;
    }

    $rowMax = max($matrix[$y]);
    if ($matrix[$y][$x] != $rowMax) {
        return false;
    }
    $colMin = $matrix[0][$x];
    for ($j = 1; $j < $height; $j++) {
        if ($matrix[$j][$x] < $colMin) {
            $colMin = $matrix[$j][$x];
        }
    }
    return $rowMax == $matrix[$y][$x] && $rowMax == $colMin;
}
